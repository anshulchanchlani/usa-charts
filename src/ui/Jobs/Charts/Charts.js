import React,{Component} from 'react';
import {ColumnChart} from 'react-chartkick'

import states from '../../.././data/states';
import TextField from '@material-ui/core/TextField';
import fetch from 'isomorphic-fetch'
import jobs from '../../../data/jobs'
import {convertArrayOfJsonObjectsToRequiredFormatForLineChart, 
        filterDataAccordingToState} from '../../.././utility/utility.js';

const NAME = "name";
const STATE = "State"
const options = {
    tooltip: { isHtml: false },    // CSS styling affects only HTML tooltips.
    legend: { position: 'right' },
    bar: { groupWidth: '90%',height:'90%' },
    
  };
const styles ={
    textField:{
        width:'80%',
        marginLeft:'10%'
    }
}
const convertedData = convertArrayOfJsonObjectsToRequiredFormatForLineChart(jobs,NAME);


export default class Charts extends Component{
    constructor(props){
        super(props)
        this.state ={jobsData:convertedData,populationData:null,filteredPopulationData:null}
        this.handleChange = this.handleChange.bind(this)
        this.filterBothCharts = this.filterBothCharts.bind(this)
    }
    handleChange(e){
        let filterText = e.target.value;
        let self = this;
        if(filterText===''){
            self.setState({jobsData:convertedData,filteredPopulationData:self.state.populationData});
            
        }else
        
        {
           this.filterBothCharts(e.target.value,NAME,STATE)
        }
        

    }
    filterBothCharts(filterText,key1,key2){
        let self =this;
        self.setState({jobsData:filterDataAccordingToState(filterText,convertedData,key1,states )})
        self.setState({filteredPopulationData:filterDataAccordingToState(filterText,self.state.populationData,key2,states)})

    }

    componentDidMount(){
         var self = this;
         this._isMounted = true;
        fetch('http://localhost:5000/getFileData')
        .then(function(response) {
          return response.json();
        })
        .then(function(myJson) {
          const convertedPopulationData = convertArrayOfJsonObjectsToRequiredFormatForLineChart(myJson,STATE);
          if(self._isMounted)
          {
              self.setState({populationData:convertedPopulationData,filteredPopulationData:convertedPopulationData});
          }
        })
        .catch(function(error){
            console.log( error);
        })
        
    }
    componentWillUnmount(){
        this._isMounted = false;
    }
    render(){
         
        return(
        <div>
        <div style={styles.textField}>
        <TextField
          id="filterField"
          label="Enter state to filter chart data"
          onChange={this.handleChange}
          fullWidth
          margin="dense"
        />
        </div>
        <ColumnChart   
            id="chartJobsVsSectors"
            xtitle="Industries" 
            ytitle="Value" 
            stacked={true}
            library={options}
            title="Population vs Jobs"
            discrete={true} 
            data={this.state.jobsData}/>
        
        {this.state.filteredPopulationData!==null && this.state.filteredPopulationData.length>0?
        <ColumnChart   
            id="chartPopulationVsYears"
            xtitle = "Year Groups"
            ytitle= "Population"
            title= "Year Group vs US Statewise Population"
            library={options}
            stacked={true}
            data={this.state.filteredPopulationData}/>:null}
        </div>
        )
    }
}