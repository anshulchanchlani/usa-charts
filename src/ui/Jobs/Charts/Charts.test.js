import React from 'react';
import Charts from './Charts';
import renderer from 'react-test-renderer';
import Enzyme, { shallow, render, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import TextField from '@material-ui/core/TextField';

Enzyme.configure({ adapter: new Adapter() });
global.shallow = shallow;
global.render = render;
global.mount = mount;
describe('Tests for Chart Component',()=>{
    const wrapper = shallow(<Charts/>)
    it('it renders correctly',()=>{
        const component = renderer.create(
            <Charts/>
        )
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    })
    
    it(' jobs vs sectors exists ',()=>{
        expect(wrapper.find('#chartJobsVsSectors').length).toEqual(1)
    })

    it(' Filter field exists',()=>{
        expect(wrapper.find('#filterField').length).toEqual(1)
    })

    it('Simulate onchange of text field',()=>{
         const mockOnChange = jest.fn();
         const event ={
             target:{
                 value:'TX'
             }
         }
         const component  = Enzyme.shallow(<TextField
            id="filterField"
            label="Enter state to filter chart data"
            onChange={mockOnChange}
            fullWidth
            margin="dense"
          />)
         component.find('#filterField').simulate('change',event);
         expect(mockOnChange).toBeCalledWith(event)
    })

})
