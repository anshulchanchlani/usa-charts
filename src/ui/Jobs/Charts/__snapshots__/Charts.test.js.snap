// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for Chart Component it renders correctly 1`] = `
<div>
  <div
    style={
      Object {
        "marginLeft": "10%",
        "width": "80%",
      }
    }
  >
    <div
      aria-describedby={undefined}
      className="MuiFormControl-root-1 MuiFormControl-marginDense-3 MuiFormControl-fullWidth-4"
    >
      <label
        className="MuiFormLabel-root-10 MuiInputLabel-root-5 MuiInputLabel-formControl-6 MuiInputLabel-animated-9 MuiInputLabel-marginDense-7"
        data-shrink={false}
        htmlFor="filterField"
      >
        Enter state to filter chart data
      </label>
      <div
        className="MuiInput-root-15 MuiInput-fullWidth-22 MuiInput-formControl-16 MuiInput-underline-19"
      >
        <input
          aria-invalid={false}
          autoComplete={undefined}
          autoFocus={undefined}
          className="MuiInput-input-23 MuiInput-inputMarginDense-24"
          defaultValue={undefined}
          disabled={false}
          id="filterField"
          name={undefined}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={undefined}
          onKeyUp={undefined}
          placeholder={undefined}
          readOnly={undefined}
          required={undefined}
          rows={undefined}
          type="text"
          value={undefined}
        />
      </div>
    </div>
  </div>
  <div
    id="chartJobsVsSectors"
    style={
      Object {
        "color": "#999",
        "fontFamily": "'Lucida Grande', 'Lucida Sans Unicode', Verdana, Arial, Helvetica, sans-serif",
        "fontSize": "14px",
        "height": "300px",
        "lineHeight": "300px",
        "textAlign": "center",
        "width": "100%",
      }
    }
  >
    Loading...
  </div>
</div>
`;
